File: ch19p620customPresentedAnimation2/ch19p620customPresentedAnimation2/ViewController2.m


#import "ViewController2.h"

@interface ViewController2 () <UIViewControllerTransitioningDelegate, UIViewControllerAnimatedTransitioning>
@property (weak, nonatomic) IBOutlet UIButton *button;

@end

@implementation ViewController2

-(id)initWithNibName:(NSString *)nib bundle:(NSBundle *)bundle {
    self = [super initWithNibName:nib bundle:bundle];
    if (self) {
        self.modalPresentationStyle = UIModalPresentationCustom;
        self.transitioningDelegate = self;
    }
    return self;
}

-(id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        self.modalPresentationStyle = UIModalPresentationCustom;
        self.transitioningDelegate = self;
    }
    return self;
}

- (IBAction)doButton:(id)sender {
    [self.presentingViewController dismissViewControllerAnimated:YES completion:nil];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.layer.borderColor = self.view.tintColor.CGColor;
    self.view.layer.borderWidth = 2;
    self.view.layer.cornerRadius = 10;
    self.view.layer.masksToBounds = YES;
    self.button.layer.borderColor = self.button.tintColor.CGColor;
    self.button.layer.borderWidth = 1;
    UIGraphicsBeginImageContextWithOptions(CGSizeMake(10,10), NO, 0);
    CGContextSetFillColorWithColor(UIGraphicsGetCurrentContext(), [UIColor colorWithWhite:0.4 alpha:1.5].CGColor);
    CGContextFillRect(UIGraphicsGetCurrentContext(), CGRectMake(0,0,10,10));
    UIImage* im = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    [self.button setBackgroundImage:im forState:UIControlStateHighlighted];
}

-(id<UIViewControllerAnimatedTransitioning>)animationControllerForPresentedController:(UIViewController *)presented presentingController:(UIViewController *)presenting sourceController:(UIViewController *)source {
    return self;
}

-(id<UIViewControllerAnimatedTransitioning>)animationControllerForDismissedController:(UIViewController *)dismissed {
    return self;
}

-(NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext {
    return 0.25;
}

-(void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext {
    UIViewController* vc1 = [transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
    UIViewController* vc2 = [transitionContext viewControllerForKey:UITransitionContextToViewControllerKey];
    UIView* con = [transitionContext containerView];
    UIView* v1 = vc1.view;
    UIView* v2 = vc2.view;
    if (v2 == self.view) {
        UIView* shadow = [[UIView alloc] initWithFrame:con.bounds];
        shadow.backgroundColor = [UIColor colorWithWhite:0.4 alpha:0.2];
        shadow.alpha = 0;
        shadow.tag = 987;
        [con addSubview: shadow];
        v2.center = CGPointMake(CGRectGetMidX(con.bounds), CGRectGetMidY(con.bounds));
        v2.autoresizingMask = UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleBottomMargin | UIViewAutoresizingFlexibleRightMargin | UIViewAutoresizingFlexibleLeftMargin;
        CGAffineTransform scale = CGAffineTransformMakeScale(1.6,1.6);
        v2.transform = CGAffineTransformConcat(scale, v2.transform);
        v2.alpha = 0;
        [con addSubview: v2];
        v1.tintAdjustmentMode = UIViewTintAdjustmentModeDimmed;
        [UIView animateWithDuration:0.25 animations:^{
            v2.alpha = 1;
            v2.transform = CGAffineTransformConcat(CGAffineTransformInvert(scale), v2.transform);
            shadow.alpha = 1;
        } completion:^(BOOL finished) {
            [transitionContext completeTransition:YES];
        }];
    } else {
        UIView* shadow = [con viewWithTag:987];
        [UIView animateWithDuration:0.25 animations:^{
            shadow.alpha = 0;
            v1.transform = CGAffineTransformScale(v1.transform,0.5,0.5);
            v1.alpha = 0;
        } completion:^(BOOL finished) {
            v2.tintAdjustmentMode = UIViewTintAdjustmentModeAutomatic;
            [transitionContext completeTransition:YES];
        }];
    }
}




@end

File: bk2ch06p324containerViewControllerStoryboard/ch19p642containerViewControllerStoryboard/ViewController2.swift
import UIKit

class ViewController2 : UIViewController {
    @IBAction func doButton(_ sender:AnyObject?) {
        self.presenting!.dismiss(animated:true, completion: nil)
    }
    override func viewDidLoad() {
        super.viewDidLoad()
        print("vc2 view did load")
    }
}

// ===== the rest of the example actually demonstrates something else entirely,
// namely a current-context presentation (with custom animation)

extension ViewController2 {
    override func awakeFromNib() {
        super.awakeFromNib()
        self.transitioningDelegate = self
        self.modalPresentationStyle = .currentContext
    }
}

extension ViewController2 : UIViewControllerTransitioningDelegate {
    @objc(animationControllerForPresentedController:presentingController:sourceController:)
    func animationController(forPresentedController presented: UIViewController, presenting: UIViewController, sourceController source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        return self
    }
}

extension ViewController2 : UIViewControllerAnimatedTransitioning {
    func transitionDuration(_ transitionContext: UIViewControllerContextTransitioning?) -> NSTimeInterval {
        return 0.8
    }
    
    func animateTransition(_ transitionContext: UIViewControllerContextTransitioning) {
        let vc2 = transitionContext.viewController(forKey:UITransitionContextToViewControllerKey)!
        
        let con = transitionContext.containerView()!
        
        let r2end = transitionContext.finalFrame(for:vc2)
        
        let v2 = transitionContext.view(forKey:UITransitionContextToViewKey)!
                
        var r2start = r2end
        r2start.origin.y -= r2start.size.height
        v2.frame = r2start
        con.addSubview(v2)
        UIView.animate(withDuration:0.8, delay: 0, usingSpringWithDamping: 0.5, initialSpringVelocity: 10, animations: {
            v2.frame = r2end
            }, completion: {
                _ in
                transitionContext.completeTransition(true)
            })

    }
}